rmqproducer/rabbitmq_producer.py

```diff
@@ -0,0 +1,390 @@
+                                       internal=self.exchange_internal)
+
+    def on_exchange_declareok(self, unused_frame):
+        """Invoked by pika when RabbitMQ has finished the Exchange.Declare RPC
+        command.
+
+        :param pika.Frame.Method unused_frame: Exchange.DeclareOk response frame
+
+        """
+        self._LOGGER.info('Exchange declared')
+        self.start_publishing()
+
+    def start_publishing(self):
+        """This method will enable delivery confirmations
+
+        """
+        self._LOGGER.info('Issuing consumer related RPC commands')
+        if self.delivery_confirmation:
+            self.enable_delivery_confirmations()
+        else:
+            self._connection.ioloop.stop()
+
+    def enable_delivery_confirmations(self):
+        """Send the Confirm.Select RPC method to RabbitMQ to enable delivery
+        confirmations on the channel. The only way to turn this off is to close
+        the channel and create a new one.
+
+        When the message is confirmed from RabbitMQ, the
+        on_delivery_confirmation method will be invoked passing in a Basic.Ack
+        or Basic.Nack method from RabbitMQ that will indicate which messages it
+        is confirming or rejecting.
+
+        """
+        self._LOGGER.info('Issuing Confirm.Select RPC command')
+        self._channel.confirm_delivery(self.on_delivery_confirmation)
+        self._connection.ioloop.stop()
+
+    def on_delivery_confirmation(self, method_frame):
+        """Invoked by pika when RabbitMQ responds to a Basic.Publish RPC
+        command, passing in either a Basic.Ack or Basic.Nack frame with
+        the delivery tag of the message that was published.
+
+        The delivery tag is an integer counter indicating the message number
+        that was sent on the channel via Basic.Publish. Here we're just doing
+        house keeping to keep track of stats and remove message numbers that
+        we expect a delivery confirmation of from the list used to keep track
+        of messages that are pending confirmation.
+
+        :param pika.frame.Method method_frame: Basic.Ack or Basic.Nack frame
+
+        """
+        confirmation_type = method_frame.method.NAME.split('.')[1].lower()
+        message_num = method_frame.method.delivery_tag
+        if confirmation_type == 'ack':
+            self._LOGGER.info('Message %i published successfully',
+                              message_num)
+            self._LOGGER.debug('The message published: %s',
+                               self._messages[message_num]['message'])
+        if confirmation_type == 'nack':
+            self._LOGGER.error('The message %i failed to publish: %s',
+                               message_num, self._messages[message_num]['message'])
+            if self.nack_callback:
+                self.nack_callback(self._messages[message_num]['message'])
+        del self._messages[message_num]
```

Comment:
No use of storing the successfully acknowledged messages, only wastage of memory